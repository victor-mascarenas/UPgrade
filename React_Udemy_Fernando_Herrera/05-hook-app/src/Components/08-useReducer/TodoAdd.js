import React, { useRef } from 'react';
import useForm from '../../hooks/useForm';

export const TodoAdd = ({addTodo}) => {
    const [{description}, inputOnChange, reset] = useForm({
        description: ''
    });
    const input = useRef();

    const formOnSubmit = (e) => {
        e.preventDefault();
        if (description.trim().length <= 1) {
            return;
        }
        addTodo({
            id: new Date().getTime(),
            desc: description,
            done: false
        });
        reset();
        input.current.focus();
    };

    return (
        <>
            <h4>Agregar ToDo</h4>
            <hr/>
            <form onSubmit={formOnSubmit}>
                <input type='text' name='description' placeholder='Comprar...' autoComplete='false' className='form-control' onChange={inputOnChange} value={description} ref={input}/>
                <button className='btn btn-outline-primary mt-1 col-12' type='submit'>Agregar</button>
            </form>
        </>
    );
};
