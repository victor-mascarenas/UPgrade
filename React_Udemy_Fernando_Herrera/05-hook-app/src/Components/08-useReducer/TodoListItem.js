import React from 'react';

export const TodoListItem = ({todo, i: index, pOnClick, buttonOnClick}) => {
    return (
        <li key={todo.id} className='list-group-item'>
            <p onClick={pOnClick.bind(null, todo.id)} className={todo.done ? 'completed' : ''}>
                {index + 1}. {todo.desc}
            </p>
            <button className='btn btn-danger' onClick={buttonOnClick.bind(null, todo.id)}>Borrar</button>
        </li>
    );
};
