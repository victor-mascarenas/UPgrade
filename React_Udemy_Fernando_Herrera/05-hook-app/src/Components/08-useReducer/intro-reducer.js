const todosInitial = [{//Estado inicial
    id: 1,
    todo: 'Comprar pan',
    done: false
}];

const todoReducer = (state = todosInitial, action) => {
    if (action) {
        if (action.type === 'agregar') {
            state = [...state, action.payload];
        }
    }
    return state;
};

let todos = todoReducer();

const newTodo = {
    id: 2,
    todo: 'Comprar leche',
    done: false
};
const action = {
    type: 'agregar',
    payload: newTodo
};
todos = todoReducer(todos, action);

console.log(todos);