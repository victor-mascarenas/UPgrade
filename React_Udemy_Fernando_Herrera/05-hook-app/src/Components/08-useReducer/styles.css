body {
    padding: 50px;
}
p {
    cursor: pointer;
    margin: 0px;
}

.complete {
    text-decoration: line-through;
}
.list-group-item {
    align-items: center;
    display: flex;
    justify-content: space-between;
}