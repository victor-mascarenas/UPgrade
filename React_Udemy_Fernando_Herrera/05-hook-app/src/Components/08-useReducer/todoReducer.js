//Delete/toggle: paylod = id
export const todoReducer = (state, action) => {
    if (action) {
        switch (action.type) {
            case 'add':
                state = [...state, action.payload];
                break;
            case 'delete':
                state = state.filter((todo) => todo.id !== action.payload);
                break;
            case 'toggle':
                state = state.map(todo =>
                    (todo.id === action.payload)
                        ? {...todo, done: !todo.done}
                        : todo
                );
                break;
            default:
                break;
        }
    }
    return state;
};
