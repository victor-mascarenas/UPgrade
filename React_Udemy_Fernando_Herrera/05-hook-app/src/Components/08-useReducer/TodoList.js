import React from 'react';
import { TodoListItem } from './TodoListItem';

export const TodoList = ({todos, pOnClick, buttonOnClick}) => {
    return (
        <ul className='list-group list-group-flush'>
            {
                todos.map((todo, i) => {
                    return <TodoListItem key={todo.id} todo={todo} i={i} pOnClick={pOnClick} buttonOnClick={buttonOnClick}/>;
                })
            }
        </ul>
    );
};
