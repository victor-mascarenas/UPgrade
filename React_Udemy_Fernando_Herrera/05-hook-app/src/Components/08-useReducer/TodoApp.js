import React, { useEffect, useReducer, useRef } from 'react';
import './styles.css';
import { todoReducer } from './todoReducer';
import useForm from '../../hooks/useForm';
import { TodoList } from './TodoList';

const init = () => {
    return JSON.parse(localStorage.getItem('todos')) || [];
};

export const TodoApp = () => {
    const [todos, dispatch]= useReducer(todoReducer, [], init);
    const [{description}, inputOnChange, reset] = useForm({
        description: ''
    });
    const input = useRef();
    useEffect(() => {
        localStorage.setItem('todos', JSON.stringify(todos));
    }, [todos]);

    const formOnSubmit = (e) => {
        e.preventDefault();
        if (description.trim().length <= 1) {
            return;
        }
        dispatch({
            type: 'add',
            payload: {
                id: new Date().getTime(),
                desc: description,
                done: false
            }
        });
        reset();
        input.current.focus();
    };
    const buttonOnClick = (id) => {
        dispatch({
            type: 'delete',
            payload: id
        });
    };
    const pOnClick = (id) => {
        dispatch({
            type: 'toggle',
            payload: id
        });
    };

    return (
        <div>
            <h1>Todo App ({todos.length})</h1>
            <hr/>
            <div className='row'>
                <div className='col-7'>
                    <TodoList todos={todos} pOnClick={pOnClick} buttonOnClick={buttonOnClick}/>
                </div>
                <div className='col-5'>
                    <h4>Agregar ToDo</h4>
                    <hr/>
                    <form onSubmit={formOnSubmit}>
                        <input type='text' name='description' placeholder='Comprar...' autoComplete='false' className='form-control' onChange={inputOnChange} value={description} ref={input}/>
                        <button className='btn btn-outline-primary mt-1 col-12' type='submit'>Agregar</button>
                    </form>
                </div>
            </div>
        </div>
    );
};
