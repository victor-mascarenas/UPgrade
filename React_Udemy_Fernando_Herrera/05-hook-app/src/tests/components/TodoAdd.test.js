import React from "react";
import {shallow, mount} from 'enzyme';
import { TodoAdd } from "../../Components/08-useReducer/TodoAdd";
import { act } from "@testing-library/react";

describe('Pruebas en TodoAdd', () => {
    const addTodoMock = jest.fn(todo => {});
    const wrapper = shallow(<TodoAdd addTodo={addTodoMock}/>);

    test('Debe de mostrarse correctamente', () => {
        expect(wrapper).toMatchSnapshot();
    });
    test('No debe de llamar addTodo', () => {
        const formSubmit = wrapper.find('form').prop('onSubmit');
        formSubmit({
            preventDefault: () => {}
        });
        expect(addTodoMock).not.toHaveBeenCalled();
    });
    test('Debe de llamar la funcion addTodo', () => {
        const value = 'Aprender Selenium';
        // mount & act for useRef call:
        const wrapper = mount(<TodoAdd addTodo={addTodoMock}/>);
        const input = wrapper.find('input');
        act(() => {
            input.simulate('change', {
                target: {
                    name: 'description',
                    value: value
                }
            });
        });
        const form = wrapper.find('form');
        act(() => {
            form.simulate('submit', {
                preventDefault: () => {}
            });
        });
        wrapper.update();
        //No mount no ref:
        /* const onChange = wrapper.find('input').prop('onChange');
        onChange({
            target: {
                name: 'description',
                value: value
            }
        });
        const onSubmit = wrapper.find('form').prop('onSubmit');
        onSubmit({
            preventDefault: () => {}
        }); */

        expect(addTodoMock).toHaveBeenCalledWith({
            id: expect.any(Number),
            desc: value,
            done: false
        });
        expect(wrapper.find('input').prop('value')).toBe('');
    });
    
});
