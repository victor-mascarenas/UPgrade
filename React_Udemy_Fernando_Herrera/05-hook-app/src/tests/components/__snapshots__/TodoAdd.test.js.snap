// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en TodoAdd Debe de mostrarse correctamente 1`] = `
<Fragment>
  <h4>
    Agregar ToDo
  </h4>
  <hr />
  <form
    onSubmit={[Function]}
  >
    <input
      autoComplete="false"
      className="form-control"
      name="description"
      onChange={[Function]}
      placeholder="Comprar..."
      type="text"
      value=""
    />
    <button
      className="btn btn-outline-primary mt-1 col-12"
      type="submit"
    >
      Agregar
    </button>
  </form>
</Fragment>
`;
